PXC-3989: Support for Keyring Components in PXC

venkatesh-prasad-v opened the pull request:
PXC-3989: Support for Keyring Components in PXC
 
https://jira.percona.com/browse/PXC-3989

**Design**
    
    1. Donor and Joiner nodes have their respective keyring components
       configured. It is the requirement for plugins as well. i.e, the
       manifest and component’s cnf file must be in-place prior to startup of
       joiner node.
    
    2. Joiner asks donor for SST.
    
    3. Donor spawns SST script and the script checks if keyring component is
       active. To get this information, the SST script will now connect to
       dono server as 'mysql.pxc.sst.user' user and queries the server's
       performance_schema.keyring_component_status table.
    
    4. Donor creates the backup using transition key, and send the
       transition key to joiner through the sst-info file.
    
    5. Transition-key encrypted backup is transferred to the Joiner
    
    6. Upon receiving the backup, the joiner node sees that transition-key
       is used and prepares the backup using the transition key received
       from the donor server.
    
    7. Joiner copies its keyring_component config file into the received backup
        directory (this is needed by PXB).
    
    8. Joiner copy-back the backup and creates a new master key and
       reencrypts the tablespace headers with the new master key.
    
    9. Once the SST is completed, the server reloads the keyring component
       to fetch the new Master Key generated by PXB.
    
       This is necessary because the server initializes keyring component
       before SST, and would have already cached the key before SST takes
       place, so the server will need to reload the component/invalidate its
       cache to learn about new MK generated by xtrabackup copy-back.

**Testing Done**

Jenkins: https://pxc.cd.percona.com/view/8.0%20parallel%20MTR/job/pxc-8.0-param-parallel-mtr/84/

Result: 

Failing tests:
1. sys_vars.wsrep_incoming_addresses_basic - Fails due to port collision
2. sys_vars.wsrep_min_log_verbosity - Fails due to port collision
3. galera_3nodes_nbo.galera_nbo_master_non_prim_failure - sporadic failure - reported https://jira.percona.com/browse/PXC-4125
4. galera_3nodes.galera_parallel_autoinc_manytrx2 - sporadic failure

Commit:
PXC-3989: Support for Keyring Components in PXC

https://jira.percona.com/browse/PXC-3989

Step-1: Revert "PXC-3107: Remove plugin dir parameter from SST parameters"
(commit fdee9f9dcfc4ccac96099b60e6ab5b71bcc86d73) and make the server to
pass the `--plugin-dir` as a parameter while invoking the SST script.

This changes is necessary for automatic detection of component status on
the joiner node.

kamil-holubicki commented on scripts/wsrep_sst_xtrabackup-v2.sh:
```diff
@@ -2015,6 +2098,24 @@ then
     rm -f "${DATA}/${IST_FILE}"
 
+    # Get keyring component status
+    get_mysqld_path
```
We are not checking the return code of `get_mysqld_path` here. Original code exited the script if mysqld was not found.

venkatesh-prasad-v replied:
You are right. I will make the script to exit if mysqld binary was not found.

venkatesh-prasad-v replied:
Done. Modified the `return` to `exit`.

kamil-holubicki commented on scripts/wsrep_sst_xtrabackup-v2.sh:
```diff
@@ -2015,6 +2098,24 @@ then
+    if [[ -z plugin_dir && -n $keyring_manifest_file ]]; then
+        wsrep_log_error "******************* FATAL ERROR ********************** "
+        wsrep_log_error "FATAL: The server is configured to use keyring, but the plugin_dir is not set."
+        wsrep_log_error "Please set the plugin_dir in the cnf file and retry."
```
Is it necessary to put it in my.cnf? We are passing it from server via cmd line parameter to sst script. So it looks like passing it via cmd line to mysqld should work.

venkatesh-prasad-v replied:
Thanks for catching this. This is no more a requirement now since we pass the plugin-dir as a parameter to SST script. I will remove it.

venkatesh-prasad-v replied:
Done

kamil-holubicki commented on scripts/wsrep_sst_xtrabackup-v2.sh:
```diff
@@ -2213,15 +2326,28 @@ then
 
+        if [ $keyring_component_enabled -eq 1 ]; then
+            # Handle joiner side component specific cnf file here
+            if [ ! -e $keyring_config_file ]; then
```
Is it possible? we set above keyring_component_enabled=1 if there is mainfiest and config

venkatesh-prasad-v replied:
That's true. I have removed this check.

venkatesh-prasad-v replied:
Done

kamil-holubicki commented on scripts/wsrep_sst_xtrabackup-v2.sh:
```diff
@@ -1008,6 +1014,63 @@ cleanup_donor()
+            fi
+        fi
+
+    wsrep_log_debug "Using manifest file: $manifest_file"
```
indentation

venkatesh-prasad-v replied:
Done

kamil-holubicki reviewed (commented):
Commit message: "dono server" -> donor server

Is it possible to introduce MTR tests?

venkatesh-prasad-v commented:
@kamil-holubicki Yes. MTR work is still in progress and I will be adding the testcases soon.

dutow commented on scripts/wsrep_sst_xtrabackup-v2.sh:
```diff
@@ -1008,6 +1014,62 @@ cleanup_donor()
+    # Get keyring manifest file path
+
+    if [ -e $mysqld_dir/$binary.my ]; then
+        local local_manifest=$(cat $mysqld_dir/$binary.my | tr "\n" ' ' | grep "read_local_manifest" | cut -d: -f2 | grep -oF "true")
```
```
{ "foo": true,
  "read_local_manifest": false
}
```
will result in true.

venkatesh-prasad-v replied:
Fixed.

dutow commented on scripts/wsrep_sst_xtrabackup-v2.sh:
```diff
@@ -1008,6 +1014,62 @@ cleanup_donor()
+
+        # Get keyring config path
+        if [ -e $plugin_dir/$keyring_component_type.cnf ]; then
+            local local_config=$(cat $plugin_dir/$keyring_component_type.cnf | tr "\n" ' ' | grep "read_local_config" | cut -d: -f2 | grep -oF "true")
```
Same issue.

venkatesh-prasad-v replied:
Fixed

venkatesh-prasad-v commented:
Jenkins: https://pxc.cd.percona.com/view/8.0%20parallel%20MTR/job/pxc-8.0-param-parallel-mtr/88/testReport/

kamil-holubicki reviewed (commented):
As we talked about on Slack, please add one more scenario when GCache and WS Cache encryptions are enabled.
Everything else looks fine. Great job.

kamil-holubicki commented on mysql-test/suite/galera/t/pxc_component_keyring_file.test:
```diff
@@ -53,10 +63,49 @@
+  --source include/assert_grep.inc
+
+  # Check that GCache ring buffer file content is encrypted on node_2
   --connection node_1
```
node_2

venkatesh-prasad-v replied:
Fixed

kamil-holubicki commented on mysql-test/suite/galera/t/pxc_component_keyring_file.test:
```diff
@@ -53,10 +63,49 @@
+  --source include/assert_grep.inc
+
+  # Check that GCache ring buffer file content is not encrypted on node_2
+  --connection node_1
```
node_2

venkatesh-prasad-v replied:
Fixed

kamil-holubicki commented on mysql-test/suite/galera/t/pxc_component_keyring_file.test:
```diff
@@ -0,0 +1,136 @@
+  --let $restart_parameters = "restart"
+  --source suite/galera/include/restart_cluster_with_sst.inc
+
+  # Check that GCache ring buffer file content is not encrypted on node_1
```
We should switch back to node_1

venkatesh-prasad-v replied:
We are already on `node_1`. The inc file `suite/galera/include/restart_cluster_with_sst.inc` switches to connection `node_1` in the end.

kamil-holubicki replied:
I would add it anyway, to avoid the confusion

venkatesh-prasad-v replied:
Done. Edited.

kamil-holubicki reviewed (approved):
lgtm

Commit:
PXC-3989: Support for Keyring Components in PXC

https://jira.percona.com/browse/PXC-3989

Step-2: This commit implements the support for keyring components in PXC.

Design:

1. Donor and Joiner nodes have their respective keyring components
   configured. It is the requirement for plugins as well. i.e, the
   manifest and component’s cnf file must be in-place prior to startup of
   joiner node.

2. Joiner asks donor for SST.

3. Donor spawns SST script and the script checks if keyring component is
   active. To get this information, the SST script will now connect to
   dono server as 'mysql.pxc.sst.user' user and queries the server's
   performance_schema.keyring_component_status table.

4. Donor creates the backup using transition key, and send the
   transition key to joiner through the sst-info file.

5. Transition-key encrypted backup is transferred to the Joiner

6. Upon receiving the backup, the joiner node sees that transition-key
   is used and prepares the backup using the transition key received
   from the donor server.

7. Joiner copies its keyring_component config file into the received
   backup directory (this is needed by PXB).

8. Joiner copy-back the backup and creates a new master key and
   reencrypts the tablespace headers with the new master key.

9. Once the SST is completed, the server reloads the keyring component
   to fetch the new Master Key generated by PXB.

   This is necessary because the server initializes keyring component
   before SST, and would have already cached the key before SST takes
   place, so the server will need to reload the component/invalidate its
   cache to learn about new MK generated by xtrabackup copy-back.
